### ISSUE TITLE
Generic dataclass fields don't work anymore

### ISSUE THREAD

**Issue by vberlier** [microsoft/pyright#2366]
In a recent release Pyright started reporting errors for code like this:

```python
@dataclass
class Foo(Generic[T]):
    value: Union[str, T]

class Bar(Foo[int]):
    pass

print(Bar(123))  # error
```

```
error: Could not bind method "__init__" because "Bar" is not assignable to parameter "self"
    Type "Foo[T@Foo]" cannot be assigned to type "Bar" (reportGeneralTypeIssues)
error: Could not bind method "__new__" because "Type[Bar]" is not assignable to parameter "cls"
    Type "Foo[T@Foo]" cannot be assigned to type "Bar" (reportGeneralTypeIssues)
```

I'm using version `1.1.174`.

**Comment by erictraut** [comment]
Thanks for the bug report, and apologies for the regression. 

I've fixed the problem and added a test case inspired by your sample.

**Comment by erictraut** [comment]
I've published pyright 1.1.175 as a "hot fix" release that addresses this regression. Sorry for any inconvenience.

**Comment by vberlier** [comment]
Thanks for the quick fix!
